src/nvim/eval.c

```diff
@@ -13277,6 +13281,71 @@ static void f_searchpos(typval_T *argvars, typval_T *rettv)
     list_append_number(rettv->vval.v_list, (varnumber_T)n);
 }
 
+/// "serverlist()" function
+static void f_serverlist(typval_T *argvars, typval_T *rettv)
+{
+  size_t n;
+  char **addrs = server_address_list(&n);
+
+  // Copy addrs into a linked list.
+  rettv->v_type = VAR_LIST;
+  rettv->vval.v_list = list_alloc();
+  rettv->vval.v_list->lv_refcount++;
+  for (size_t i = 0; i < n; i++) {
+    listitem_T *li = listitem_alloc();
+    li->li_tv.v_type = VAR_STRING;
+    li->li_tv.v_lock = 0;
+    li->li_tv.vval.v_string = (char_u *) addrs[i];
+    list_append(rettv->vval.v_list, li);
+  }
+  free(addrs);
+}
+
+/// "serverlisten()" function
+static void f_serverlisten(typval_T *argvars, typval_T *rettv)
+{
+  rettv->v_type = VAR_STRING;
+  rettv->vval.v_string = NULL;  // Will hold the address of the new server.
+
+  if (check_restricted() || check_secure()) {
+    return;
+  }
+
+  // If the user supplied an address, use it, otherwise use a temp.
+  if (argvars[0].v_type != VAR_UNKNOWN) {
+    if (argvars[0].v_type != VAR_STRING) {
+      EMSG(_(e_invarg));
+      return;
+    } else {
+      rettv->vval.v_string = vim_strsave(get_tv_string(argvars));
+    }
+  } else {
+    rettv->vval.v_string = vim_tempname();
+  }
+
+  bool result;
+  if ((result = server_start((char *) rettv->vval.v_string)) != 0) {
+    EMSG2("Failed to start server: %s", uv_strerror(result));
+  }
```

Comment:
Agreed, the line's a little busy. Also, noticed that `result` should be an `int`.